### PR TITLE
HSEARCH-3068 Restore support for "simulated" transaction using a Flush listener

### PR DESCRIPTION
**Pull request by fax4ever**
https://hibernate.atlassian.net/browse/HSEARCH-3068

Based on #2054

### CODE REVIEW

File: util/common/src/main/java/org/hibernate/search/util/common/concurrent/ConcurrentReferenceHashMap.java [hunk 1]
```diff
@@ -0,0 +1,1922 @@
...
+ * This class is a member of the <a href="{@docRoot}/../technotes/guides/collections/index.html">
+ * Java Collections Framework</a>.
+ *
+ * @param <K> the type of keys maintained by this map
+ * @param <V> the type of mapped values
+ *
+ * @author Doug Lea
+ * @author Jason T. Greene
+ */
+class ConcurrentReferenceHashMap<K, V> extends AbstractMap<K, V>
```
Comment (yrodiere) [inline util/common/src/main/java/org/hibernate/search/util/common/concurrent/ConcurrentReferenceHashMap.java]: Another reason not to use a `Synchronization` map: this class has proven problematic in the past and I've worked hard to get rid of it in Search 6, be it only because it's too complex for its own good. Please don't restore it.

File: util/common/src/main/java/org/hibernate/search/util/common/concurrent/ConcurrentReferenceHashMap.java [hunk 2]
```diff
@@ -20,6 +20,7 @@
 import java.util.AbstractCollection;
 import java.util.AbstractMap;
```
Comment (yrodiere) [inline util/common/src/main/java/org/hibernate/search/util/common/concurrent/ConcurrentReferenceHashMap.java]: When you restore classes that were obviously copy/pasted from another project, please don't refactor them. It makes it very hard to track differences between projects.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java [hunk 3]
```diff
@@ -139,16 +144,15 @@ public void onPostUpdateCollection(PostCollectionUpdateEvent event) {
...
-//		Synchronization synchronization = flushSynch.get( session );
-//		if ( synchronization != null ) {
-//			//first cleanup
-//			flushSynch.remove( session );
-//			log.debug( "flush event causing index update out of transaction" );
-//			synchronization.beforeCompletion();
-//			synchronization.afterCompletion( Status.STATUS_COMMITTED );
-//		}
+		Session session = event.getSession();
+		Synchronization synchronization = flushSynch.get( session );
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java]: I don't know if it a good idea always calling execute on the work plan. In such way we could invoke execute on flush even if there is no update outside transactions on the current session.

It would nice to have if there was a way to detect if there are some out of transaction update operations pending on the current session.

Currently we rely on the weird synchronization map to detect them. If there was another way, I would happy about it.

**Comment by fax4ever** [comment]
Made all the changes here. It looks much better now.
Thanks

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java [hunk 4]
```diff
@@ -131,16 +133,32 @@ public void onPostUpdateCollection(PostCollectionUpdateEvent event) {
...
+	public void onAutoFlush(AutoFlushEvent event) throws HibernateException {
+		getCurrentWorkPlan( state.getContextProvider(), event.getSession() ).prepare();
+	}
+
+	@Override
+	public void onClear(ClearEvent event) {
+		EventSource session = event.getSession();
+		// skip the clearNotPrepared operation in case of rollback
+		if ( session.isTransactionInProgress() ) {
+			getCurrentWorkPlan( state.getContextProvider(), session ).clearNotPrepared();
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java]: I get why you originally added this, but I think it's wrong now that you support flush events outside of transactions. What happens when you change entities, clear, then flush, outside of a transaction? It shoudn't do anything, yet with this code I suspect it will update the index, or result in a LazyInitializationException, because the clear is ignored.

What was the problem with executing this code outside of a transaction, exactly?

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/session/SearchSessionWritePlanOutOfTransactionIT.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	@PortedFromSearch5(original = "TransactionTest#testTransactionCommit")
+	public void add() {
+		withinSession( sessionFactory, session -> {
+			IndexedEntity entity1 = new IndexedEntity( 1, "number1" );
+			session.setHibernateFlushMode( FlushMode.AUTO );
+			session.persist( entity1 );
+
+			// TODO HSEARCH-3068 support it
+			SubTest.expectException( () -> {
+				/*backendMock.expectWorks( IndexedEntity.INDEX_NAME )
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/session/SearchSessionWritePlanOutOfTransactionIT.java]: Ok, so you squashed the test together with the fix. That's not what I meant, but to each his own.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+		backendMock.expectAnySchema( IndexedEntity.INDEX_NAME );
+		sessionFactory = ormSetupHelper
+				.start()
+				.withProperty( AvailableSettings.ALLOW_UPDATE_OUTSIDE_TRANSACTION, true )
+				.setup( IndexedEntity.class );
+		backendMock.verifyExpectationsMet();
+	}
+
+	@Test
+	@PortedFromSearch5(original = "TransactionTest#testTransactionCommit")
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java]: This is not at all the same test as `TransactionTest#testTransactionCommit`, though, since the original was inside a transaction and this one is not.

Also, if you add `@PortedFromSearch5` annotations, please use the fully qualified name of the class, and mark the original method with `@Category(PortedToSearch6.class)`. This will help later when we'll try to port more tests from Search 5 to 6.

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java [hunk 7]
```diff
@@ -137,18 +133,18 @@ public void onPostUpdateCollection(PostCollectionUpdateEvent event) {
...
-//			synchronization.beforeCompletion();
-//			synchronization.afterCompletion( Status.STATUS_COMMITTED );
-//		}
+		HibernateOrmListenerContextProvider contextProvider = state.getContextProvider();
+		EventSource session = event.getSession();
+		PojoWorkPlan currentWorkPlan = getCurrentWorkPlan( contextProvider, session );
+		currentWorkPlan.prepare();
+
+		// flush within a transaction should trigger only the prepare phase,
+		// since the execute phase is supported to be triggered by the transaction commit
```
Comment (yrodiere) [inline mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java]: `is supported to be triggered` => `is supposed to be triggered`?

File: mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java (hunk 4 above)
Comment (fax4ever) [reply mapper/orm/src/main/java/org/hibernate/search/mapper/orm/event/impl/HibernateSearchEventListener.java]: We would have a small issue just on rollback. Since the rollback triggers a clear event out of transaction. The issue is just about the creation of a work plan for nothing,
But we can change the predicate into:
``` java
// skip the clearNotPrepared operation in case of rollback
if ( !TransactionStatus.ROLLED_BACK.equals( session.getTransaction().getStatus() ) ) {
  getCurrentWorkPlan( state.getContextProvider(), session ).clearNotPrepared();
}
```

And I have a **change entities, clear, then flush, outside of a transaction** test to add

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java (hunk 6 above)
Comment (fax4ever) [reply integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java]: You're right tests are not equals. So I removed the annotation from there.

**Comment by fax4ever** [comment]
Thanks @yrodiere. Rebased and made the changes here.

File: integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java [hunk 8]
```diff
@@ -64,6 +64,30 @@ public void add() {
...
 	}
 
+	@Test
+	public void clear() {
+		withinSession( sessionFactory, session -> {
+			SearchSessionWritePlan writePlan = Search.session( session ).writePlan();
+
+			IndexedEntity entity1 = new IndexedEntity( 1, "number1" );
+			// working directly on the work plan to add works immediately (not at flush time!)
+			writePlan.addOrUpdate( entity1 );
```
Comment (yrodiere) [inline integrationtest/mapper/orm/src/test/java/org/hibernate/search/integrationtest/mapper/orm/automaticindexing/AutomaticIndexingOutOfTransactionIT.java]: Right, that's definitely an edge case. But it's nice to have it covered :)

**Comment by fax4ever** [comment]
Thanks!
